### PR TITLE
Discourse integration

### PR DESCRIPTION
**Pull request by JKarlavige**
## Description & motivation

Asana task: https://app.asana.com/0/1200099998847559/1202844016314445/f

This adds a new `DiscourseFeed` component to be used throughout the site. This component uses the Discourse API to display topics on docs pages. A new `/forum` page is created to display different sections of Discourse topics via the new component.

This also includes an example of how the component can be used on any docs pages.

I will be doing the website work within this PR, and will open a separate PR to handle the Algolia search side of this task. I expect the Algolia work to take longer so no reason to hold up this work.

## Preview
Forum page: https://deploy-preview-2004--docs-getdbt-com.netlify.app/forum

## TODOs
- [x] Remove example use of Discourse component from Jinja docs page

### CODE REVIEW

File: website/docs/faqs/Jinja/which-jinja-docs.md
Comment (joellabes) [inline website/docs/faqs/Jinja/which-jinja-docs.md]: Could the top margin be baked into the default template? I suspect that it's something we'll always want.

File: website/functions/get-discourse-posts.js
Comment (joellabes) [inline website/functions/get-discourse-posts.js]: I assume this is copy-paste from the events feed
```suggestion
    return await returnResponse(500, { error: 'Unable to get topics from Discourse.'})
```

File: website/functions/get-discourse-posts.js
Comment (joellabes) [inline website/functions/get-discourse-posts.js]:
```suggestion
  const inStringValues = ['title', 'likes', 'personal', 'messages', 'seen', 'unseen', 'posted', 'created', 'watching', 'tracking', 'bookmarks', 'assigned', 'unassigned', 'first', 'pinned', 'wiki']
```

Comment (joellabes) [reply website/functions/get-discourse-posts.js]: I don't know which user this API key is associated with, but it's probably unnecessary to support things like personal, seen, unseen, created as they're user-specific. I guess they're not doing any harm? 

If we wanted to cut it down, I think the only ones that are generic are `title, first, pinned, wiki`.

File: website/functions/get-discourse-posts.js
Comment (joellabes) [inline website/functions/get-discourse-posts.js]: Does MDX allow duplicate keys? You're allowed to do something like [`status:open status:solved`](https://discourse.getdbt.com/search?q=status%3Aopen%20status%3Asolved), but this feels like it would break: 

`<DiscourseFeed category='help' status='solved' status='open' />`

It's not likely that we'd do this very often, so having a string instead of an array as the main input type is more ergonomic. Can you pass arrays into MDX if needed (`status=['solved', 'open']`)? Or would we have to comma-separate them and you split them out?

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]:
```suggestion
        <p>Unable to load forum posts at this time.</p>
```

I think we'll de-emphasise the platform we're using and lean into it just being the Forum (in the UI at least; the code can still be specific)

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: As the default, it would be good for the CTA to be `Ask the Community` which goes to `https://discourse.getdbt.com/new-topic?category=XXXX&tags=comma,separated`

Note that in search you can either use `,` in tags (for any of) and `+` (for all of), so you'd need to replace +s with ,s here too

https://meta.discourse.org/t/how-to-create-a-new-topic-link-with-pre-filled-information/28074

Then we could override it on the forum homepage to be the Read More ones you currently have. (Side note: are we allowed to set them to null/undefined to cause no button to be shown?)

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: Can we have relative periods here? I was thinking we'd do something like last 30 days for the help ones, and last 90 for discussions.

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: As they currently look, there's not a lot of social proof that the answers you click through to are going to be accurate. Would it be overkill to put a ✅ with `Solved` hover text in the front of solved topics? (I think it's called `has_accepted_answer` or something in the response body)

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: If you wanted to get the username without a second API call, you can get it from the `posts` part of the search response. 
image

You'll want to only take it if the `post_number` is 1. When we're searching for topic-wide metadata (e.g. tag, category, solved status) this is fine - that'll always be the one you're given. It's only if you search for specific text that you might get a different post - e.g. this one, where you get the first post in a topic that contains the search text. We don't intend to do that for any of our embeds AFAIK. 
image

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: Arguably this could be `{posts_count - 1}` replies. Not attached to this. 

Side note: `1 comments` is my biggest UI pet peeve. (Very attached to this 😅)

**Review by joellabes — CHANGES_REQUESTED** [review]
This is looking so great @JKarlavige 🤩😍

A handful of comments, hopefully nothing too arduous!

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: Can we default this to `'help'` so we don't have to specify it everywhere? That's going to be the overwhelming majority of embeds.

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: Likewise here, let's have status:solved as the default and override it where needed. 

OR, it might be easier to have a focused `DiscourseHelpFeed` which ultimately creates a `DiscourseFeed` but we don't have to put defaults into the main constructor. Is that practical?

**Review by joellabes — COMMENTED** [review]
Was just writing a comment in Asana about how we only have to specify the tags we want in help mode, and then thought "hang on a minute, is that accurate?"

File: website/functions/get-discourse-posts.js
Comment (JKarlavige) [reply website/functions/get-discourse-posts.js]: Yeah that's a good point, i'll adjust this to support multiple values for the same property.

File: website/functions/get-discourse-posts.js
Comment (JKarlavige) [reply website/functions/get-discourse-posts.js]: Yeah it doesn't hurt having those on our end as it only ensures the values will work with Discourse. But if you would like to prevent those other values from being used, we can remove them from the validation here. Will leave that up to you!

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: Updated the default text & link for the CTA. If the `category` or `tags` are passed in, it will append them to the link url by default. If a custom `link_href` is passed in, it will override the default.

For the tags, since it's a string with comma (or +) separated values, passing a string such as `<DiscourseFeed  tags='jinja,testing' />` will work as i'm passing the tags string directly into the query.

Also added a separate `hide_cta` prop to control whether the cta should be hidden.

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: Love this

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: Does this refer to the `after` property? If a different category is passed in other than `help` or `discussions` should we exclude the `after` date, or should we set a default number of days back?

I pushed an update to set a relative `after` date if the category is `help` or `discussions`, otherwise it excludes the `after` date altogether for now.

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: Set the default status to `solved`. We can create a more specific component which uses the original component with specific props if you'd like. If you feel like it would be easier for the team to work with we can get it setup. I would just want to go over all the settings that more specific component would need.

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: ahhhh good find! Will get this in there

File: website/functions/get-discourse-posts.js
Comment (JKarlavige) [reply website/functions/get-discourse-posts.js]: Updated this to support arrays for the `status` property. The docs page example shows how multiple statuses can be passed in: https://github.com/dbt-labs/docs.getdbt.com/blob/2590c0049fcdd15eb0005f10cd4c625f799d2528/website/docs/faqs/Jinja/which-jinja-docs.md

`status` can now support string: `status='open'` or arrays: `status={['open', 'public']}` when being passed into the component

**Comment by JKarlavige** [comment]
@joellabes marked comments as resolved for which i pushed updates for. I have a few open questions/clarifications for the others!

File: website/functions/get-discourse-posts.js
Comment (joellabes) [reply website/functions/get-discourse-posts.js]: I think let's remove them so that it's clear to future users which ones they can realistically use

File: website/src/components/discourse/index.js
Comment (joellabes) [reply website/src/components/discourse/index.js]: Yes I meant it for after - will have a look at the new code

**Comment by lfracalossiFA** [comment]
Hi @JKarlavige,
Is the "-" between the thread title and the byline by design? I don't love it. But I also don't know if this is something that was requested!

Screen Shot 2022-09-19 at 4 40 00 PM

**Comment by lfracalossiFA** [comment]
The fonts also seem small to me on the list items on /forum. (I think it's the overall docs font face; I feel like it skews small). Any objection to bumping those up to 18px? I'd also suggest making the title bold, eg:

Screen Shot 2022-09-19 at 4 47 38 PM

File: website/src/components/discourse/styles.module.css
Comment (lfracalossiFA) [inline website/src/components/discourse/styles.module.css:26]: I mentioned this in the discussion section of this PR, but I think we should bump up the font here to 18px, and make the `a` bold. 

(Maybe just for /forum? Not sure it would be appropriate in the in-page component, because it might distract from the doc content itself).

File: website/src/components/discourse/styles.module.css
Comment (lfracalossiFA) [inline website/src/components/discourse/styles.module.css]: I feel like the font size and weight are inconsistent with buttons like this elsewhere on the site? (Compare to homepage). But maybe there is a different button you based this off?

**Comment by lfracalossiFA** [comment]
It looks like the meta description is "Description here." Any way to customize that? 

Screen Shot 2022-09-19 at 5 18 22 PM

File: website/src/components/discourse/index.js
Comment (lfracalossiFA) [inline website/src/components/discourse/index.js:101]: 👍

File: website/src/pages/forum.js
Comment (lfracalossiFA) [inline website/src/pages/forum.js]: I saw the conversation about the `after` attribute with Joel above -- will we want a date like this to be hard-coded here? Or is this just an example?

**Review by lfracalossiFA — DISMISSED** [review]
Overall, this is really looking good -- will add great value to our pages. I have a few questions as well as requests re: styling/UI. (I didn't see a comp anywhere, so feel free to ignore my suggestions if they contradict any stakeholder requests).

File: website/src/components/discourse/index.js
Comment (joellabes) [reply website/src/components/discourse/index.js]: I had originally thought it might be another prop on the DiscourseFeed, but I don't mind how you've done it here! 

Everything that isn't `help` can be 90 days - we might wind up reconsidering this at some point but messing around with the params is the sort of thing I can change unassisted 🎉

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js:124]: Also on the social proof front, it'd be nice to pull through the `like_count` from the posts array. I think this will have more purpose on the Discussion ones as I doubt people are going to click like on the OP of a support request, but if they do we can show them there too! 

I don't know how you render them - it might get messy to have `How do I do computers? by joellabes, 1 reply, 3 likes` but if we don't have design time then that might just have to do for now.

File: website/src/pages/forum.js
Comment (joellabes) [reply website/src/pages/forum.js]: I doubt we'd ever hardcode a date in here, and instead would just use the relative calculation 👍

File: website/src/pages/forum.js
Comment (joellabes) [inline website/src/pages/forum.js]: discussions and show and tell don't have Mark as Solved enabled, so the status is redundant here

**Review by joellabes — CHANGES_REQUESTED** [review]
Yeah this is all basically done from my perspective! Handful of tweaks but then we're good to go

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: Isn't it more standard to have positive flags, so it'd be `show_cta = true`? Non-blocking

File: website/src/components/discourse/index.js
Comment (joellabes) [reply website/src/components/discourse/index.js]: Yeah I think it would be good to make a focused one. The default props: 
```javascript
order = 'latest_topic',
status = 'solved',
category = 'help',
link_text = 'Ask the Community',
link_href = // the same URL builder you've already got and I don't want to copy-paste
hide_cta = 'false'
```

Then on the main `DiscourseFeed`, everything would be undefined except for `hide_cta` which would be true. (Or, if you [changed it to be `show_cta`](https://github.com/dbt-labs/docs.getdbt.com/pull/2004#discussion_r974851260), then _everything_ could be undefined because `undefined` is falsy right?)

File: website/src/pages/forum.js
Comment (lfracalossiFA) [reply website/src/pages/forum.js]: @joellabes That's why I'm wondering about the `after='2022-06-14'` on Cool Problem Solving...

File: website/src/components/discourse/styles.module.css
Comment (JKarlavige) [reply website/src/components/discourse/styles.module.css]: Hmmmm, this button is the same one used on the homepage cards, i'm using the same `button button--primary` classes. The only thing I did was reduce the font-size a bit as i felt it was too large. Is there anything specific you think we should change with the buttons for this component?

File: website/src/components/discourse/styles.module.css
Comment (JKarlavige) [reply website/src/components/discourse/styles.module.css:26]: I bumped the font to `18px` and make it bold. I'm not crazy about having bold text in one place where the component is used, and not bold in another spot. We'd have to add another prop to control that, and there's already 18 potential props that can be used. 😅 Maybe `font-weight: 600` is a good in-between? I'll have that in my next push

**Comment by JKarlavige** [comment]
> It looks like the meta description is "Description here." Any way to customize that?
> 
> Screen Shot 2022-09-19 at 5 18 22 PM

cc @joellabes, I added title and description meta tags to the /forum page. What should the meta title and description be for the page?

File: website/src/pages/forum.js
Comment (JKarlavige) [reply website/src/pages/forum.js]: Removed the `after` property for the Cool Problem Solving component.

File: website/src/components/discourse/styles.module.css
Comment (lfracalossiFA) [reply website/src/components/discourse/styles.module.css]: > Is there anything specific you think we should change with the buttons for this component?

Hm, I guess not. From a UX perspective it's just as usable.

File: website/src/components/discourse/styles.module.css
Comment (lfracalossiFA) [reply website/src/components/discourse/styles.module.css:26]: Works for me, thanks!

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js:124]: Here's an example of how it looks with the like count added. Not many have the like count so not sure if any would show up in the preview, so adding it here. I had to set the `after` prop to `2022-01-01` on the first component within the /forum page to get it to show up.

`<DiscourseFeed title='Open topics' category='help' status='unsolved'  link_text='See open topics' link_href='https://discourse.getdbt.com/c/help/19' show_cta={true} after='2022-01-01' />`

image

**Comment by JKarlavige** [comment]
@joellabes All updates in place and ready for re-review. @lfracalossiFA i'll hold off on requesting a re-review from you until the meta fields are in place.

File: website/src/components/discourse/index.js
Comment (joellabes) [reply website/src/components/discourse/index.js:124]: Works for me!

File: website/src/pages/forum.js
Comment (joellabes) [inline website/src/pages/forum.js:6]: Is this a copy-paste hangover too? And does it matter?

**Review by joellabes — APPROVED** [review]
😻 😻 😻

File: website/src/pages/forum.js
Comment (joellabes) [inline website/src/pages/forum.js]: @JKarlavige How about "Recent interesting discussions from across the dbt Forum" for now

File: website/src/components/discourse/index.js
Comment (joellabes) [inline website/src/components/discourse/index.js]: The slug is human-readable, but the actual important part of the link is the topic ID. 

https://discourse.getdbt.com/t/compare-two-models-schemas will redirect to https://discourse.getdbt.com/t/compare-two-models-schemas/3868

https://discourse.getdbt.com/t/fjhsbjhfbshjf/3868 will also get to the right place. 

I don't think this is _likely_ to break since we're using search results from a few seconds ago and the slug probably won't have been changed by a topic rename, but it's worth doing it the proper-er way.

**Review by joellabes — CHANGES_REQUESTED** [review]

File: website/src/components/discourse/index.js
Comment (joellabes) [reply website/src/components/discourse/index.js]: (and _THEN_ I promise I'm done with feedback 😅)

File: website/src/pages/forum.js
Comment (JKarlavige) [reply website/src/pages/forum.js:6]: Since we added a second `DiscourseHelpFeed` component (which i added directly in the original component file), I had to change the way the components were exported. It was originally set to export 1 default component: `export default`, but since the file now has 2 components I changed to named exports `export const <variable-name>`. This allows the imports to be destructured with `import { DiscourseFeed, DiscourseHelpFeed }` as needed

File: website/src/components/discourse/index.js
Comment (JKarlavige) [reply website/src/components/discourse/index.js]: good call, just updated!

**Comment by JKarlavige** [comment]
All updates in place and re-requested review. Documentation for using this component has been put together and is available here: https://www.notion.so/dbtlabs/Discourse-Component-Documentation-7057e62ee25b4536b84e676ca1c0e892

**Review by joellabes — APPROVED** [review]

**Comment by JKarlavige** [comment]
- Tests added and pass successfully
- Updated to current
- Removed example from Jinja docs page
- Added /forum page to Learn dropdown in nav (i lied, we reverted the top nav PR until restructuring is complete.)
- This is ready for lift-off 🚀 

image

**Comment by JKarlavige** [comment]
@joellabes note - this will be absent from the navigation until the new top nav is ready to be taken live.

**Pull request merged**
